### PR TITLE
Particles and some minor adjustments

### PR DESCRIPTION
**Pull request by IchHabeHunger54**
This PR adds the particles and visual effects Ars Magica 2 was famous for into the mod. Specifically, it:

- Adds various new particle types with textures from the old mod, some of which were reworked into the JAPPA style
- Adds the `AMParticle` class that can take any other particle's appearance
- Adds a particle controller system that allows for complex particle movements to be defined
- Adds a particle spawning system that runs on the client upon receiving a packet from the server. The server does not spawn the particles itself, making the system suited for multiplayer
- Adds particle spawners for all components that had particle spawning in the old mod, staying as close to the old design as reasonably possible
- Replaces the old particles on some of the entities (projectile, wall, wave, zone, blizzard, firestorm, falling star, mana vortex) with proper new particles
- Integrates with the Color modifier (see #355 )

It also does some other minor adjustments:

- Changes the beam shape's color depending on the spell's primary affinity
- Changes some implementation details on some of the spell parts
- Fixes `SkillHelper` to not crash when leaving the end (we had like 5 issues like this now, I'm growing tired)
- Replaces the `LivingEntity::canBreatheUnderwater` call in `AMSpellParts` with a call to the new fluid system
- Adds missing loot tables for the three inlay blocks
- Adds new API methods for getting the spell item stack from an entity or from an interaction hand
- Removes `AMUtil.nextDouble` and replaces its uages with direct `RandomSource#nextDouble` calls
- Removes various overrides of `Entity#isPushable`, courtesy of IntelliJ's new "method is identical to supermethod" inspection
- Removes the `ItemSupplier` implementation on various entities (it was never actually needed for anything except `Projectile`)

This is a draft PR for now, so that people can review it. It will be marked as ready once all particle effects have actually been implemented.

### CODE REVIEW

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/spell/shape/Beam.java [hunk 1]
```diff
@@ -37,4 +37,8 @@ public boolean needsPrecedingShape() {
     public boolean needsToComeFirst() {
         return true;
     }
+
+    public static int getBeamColor(ISpell spell, List<ISpellModifier> modifiers) {
+        return spell.primaryAffinity().color(); //TODO color modifier
+    }
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/spell/shape/Beam.java]: why is this here?

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/spell/shape/Beam.java (hunk 1 above)
Comment (IchHabeHunger54) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/common/spell/shape/Beam.java]: leftover from before the color modifier was merged

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java [hunk 2]
```diff
@@ -255,28 +255,58 @@ private void sync() {
         menu.sendDataToServer(nameBar.getValue().equals(NAME_LABEL.getString()) ? null : Component.literal(nameBar.getValue()), spellGrammarArea.getAll().stream().map(e -> e.getPart().getId()).toList(), shapeGroupArea.getShapeGroupData(), compileAdditionalData());
     }
 
+    private void onPartClicked(SpellPartDraggable part, int button) {
+        ISpellPart spellPart = part.getPart();
+        if (spellPart == AMSpellParts.COLOR.get() && minecraft.options.keyUse.isActiveAndMatches(InputConstants.Type.MOUSE.getOrCreate(button))) {
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java]: can we please not use keybinds in the gui

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/model/block/AltarCoreModel.java [hunk 3]
```diff
@@ -12,13 +12,16 @@
...
 import net.minecraftforge.client.model.data.ModelData;
+import net.minecraftforge.fml.ModList;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class AltarCoreModel extends BakedModelWrapper<BakedModel> {
+    private static final boolean IS_SODIUM_PRESENT = ModList.get().isLoaded("embeddium") || ModList.get().isLoaded("rubidium");
```
Comment (Minecraftschurli) [inline src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/model/block/AltarCoreModel.java:23]: eeeww

**Review by Minecraftschurli — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java (hunk 2 above)
Comment (IchHabeHunger54) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java]: why not? we are already on the client

Comment (Minecraftschurli) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java]: bad practice, if someone has the use key on the keyboard this code breaks

Comment (IchHabeHunger54) [reply src/main/java/com/github/minecraftschurlimods/arsmagicalegacy/client/gui/inscriptiontable/InscriptionTableScreen.java]: The use key only handles in-world use of items, not GUI interactions. Minecraft doesn't support using GUIs without a mouse. See `AbstractContainerScreen#mouseClicked` (which is where I got that piece of code from).

**Review by Minecraftschurli — APPROVED** [review]
If it is working then :shipit:
